feat(zabbix-proxy): add support for persistent volume

tomsozolins opened the pull request:
Add persistent volume support for zabbix-proxy

tomsozolins committed:
feat(zabbix-proxy): add support for persistent volume

tomsozolins committed:
feat(zabbix-proxy): add support for persistent volume

tomsozolins committed:
feat(zabbix-proxy): add support for persistent volume

tomsozolins committed:
fix(zabbix-proxy): change volume ownership to zabbix user

tomsozolins committed:
fix(zabbix-proxy): change volume ownership to zabbix user

tomsozolins commented:
Zabbix process didn't have access to the volume as it had root ownership. Updating ownership to zabbix user id 1997 fixed the problem.

tomsozolins committed:
docs(zabbix-proxy): rename postgres to database

tomsozolins committed:
fix(zabbix-proxy): move fsgroup to proper spec section

tomsozolins committed:
fix(zabbix-proxy): move securitycontext to proper spec section

aeciopires commented:
Hi, @tomsozolins!

Thank you very much for your contribution. I really appreciate this initiative.

Sorry for the delay in reviewing your Pull Request. I was on vacation. Tomorrow I will take the time to review and test the changes. Have a good week.

CC: @sa-ChristianAnton for your information.

aeciopires commented:
Hi @tomsozolins!

I understand your contribution, but I think that parameters 

```
zabbixproxy:
  # -- additional volumeMounts to the zabbix proxy container
  extraVolumeMounts: []
  # -- additional volumes to make available to the zabbix proxy pod
  extraVolumes: [] 
```
  
I understood your contribution, but I think that parameters have same result and are more simple. Do you agree?
Could you test this parameters? 

If works, I think this Pull Request can be closed without merge.

The same idea works to zabbixserver, zabbixagent, zabbixweb.

CC: @sa-ChristianAnton

tomsozolins commented:
Hi! @aeciopires what about the volumeClaimTemplates? Can it be specified in extra fields?

aeciopires commented:
Hi @tomsozolins !

I agree with the idea of ​​securityContext having the option to be localized and with the idea of ​​changing the implementation to support extraVolumeClaimTemplate... but looking again at values.yaml... we have the extraPodSpecs parameter which makes the code generic enough to add any settings. Does this work for you?

```
zabbixproxy:
  # -- additional specifications to the zabbix proxy pod
  extraPodSpecs: {}
```

The same idea works to zabbixserver, zabbixservice, zabbixweb.

If it doesn't work, could you change the PR to add securityContext and extraVolumeClaimTemplate support for zabbixProxy, zabbixServer, zabbixWeb, zabbixWebservice?

aeciopires commented:
Hi @tomsozolins!

I'm sorry!

The extraPodSpecs won't work for you because it is at the pod level and the extraVolumeClaim would go into the Spec at the StateFulSet level (one level higher). Please disregard what I said earlier about the parameter.

Could you change the PR implementation to consider the extraVolumeClaimTemplate correctly (making it very generic similar to the other extra parameters) and leave the securityContext localized, overriding the global? Example: if the localized securityContext exists, use it, otherwise use the global one.

tomsozolins commented:
@aeciopires Ok, i will look into this later.

tomsozolins committed:
fix(zabbixproxy): add extraVolumeClaimTemplate

Cleanup volume mount templates - extra sections is enough for use case

tomsozolins committed:
fix(zabbixproxy): add extraVolumeClaimTemplate

Cleanup redundant values - extra values is enough for use case

tomsozolins commented:
@aeciopires Hi! I removed all the extra stuff i added previously and added only extraVolumeClaimTemplate. 
This is enough to add persistence for zabbixproxy:
`
extraVolumeMounts:
    - mountPath: /var/lib/zabbix/db_data
      name: zabbixproxy-data
extraPodSpecs:
    securityContext:
        fsGroup: 1997
extraVolumeClaimTemplate:
    - metadata:
        name: zabbixproxy-data
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
        storageClassName: gp3
`

aeciopires commented on templates/statefulset-zabbix-proxy.yaml:
```diff
@@ -147,4 +147,6 @@ spec:
       volumes:
       {{- toYaml . | nindent 8 }}
       {{- end }}
+  volumeClaimTemplates:
```
Hi @tomsozolins!

Good job here. Do you can change this block to?

```
  {{- with .Values.zabbixproxy.extraVolumeMounts }}
  volumeClaimTemplates:
    {{- toYaml .Values.zabbixproxy.extraVolumeClaimTemplate | nindent 4 }}
  {{- end }}
```

tomsozolins replied:
extraVolumeMounts is under different template section, this wouldn't work?

aeciopires replied:
Sorry @tomsozolins, error of copy/past. I fix the suggestion.

tomsozolins replied:
@aeciopires adjusted

aeciopires replied:
Hi @tomsozolins! Do you make the commit? I don't see the changes in this file.

aeciopires commented on values.yaml:
```diff
@@ -218,6 +217,8 @@ zabbixproxy:
   # -- additional specifications to the zabbix proxy pod
   extraPodSpecs: {}
+  # -- extra volumeClaimTemplate for zabbixproxy statefulset
+  # extraVolumeClaimTemplate: []
```
Hi @tomsozolins!

With my suggestion implemented in before file, you can to uncomment the new parameter.

tomsozolins replied:
@aeciopires adjusted

aeciopires reviewed (changes requested):
Hi @tomsozolins!

Great job! Thanks for your contribution.

I wroten some suggestions.

Please, do you can make other changes?

1) In *values.yaml* file change ubuntu-6.0.7 to ubuntu-6.0.8

2) In *docs/example/kind/values.yaml* file change alpine-6.2.0 to alpine-6.2.2

3) Run the *helm-docs* command following the instructions of *CONTRIBUTING.md* file to update the README.md.

Have a nice day.

CC: @sa-ChristianAnton do you can review this pull request too?

tomsozolins committed:
feat(zabbix): update app appVersion to 6.0.8

tomsozolins committed:
fix(zabbixproxy): uncomment extraVolumeClaimTemplate

tomsozolins commented on Chart.yaml:
```diff
@@ -1,7 +1,7 @@
-version: 3.1.1
+version: 3.1.2
 # LTS Zabbix version by default due to stability. See: https://www.zabbix.com/life_cycle_and_release_policy
 appVersion: 6.0.7
```
@aeciopires adjusted

tomsozolins committed:
feat(zabbix): update zabbix_image_tag to 6.0.8

tomsozolins committed:
feat(zabbix): update example zabbix_image_tag to alpine-6.2.2

tomsozolins committed:
docs(zabbix): update readme

tomsozolins committed:
docs(zabbix): update readme

aeciopires commented on README.md:
```diff
@@ -1,334 +1,26 @@
-# Helm Chart For Zabbix.
+# zabbix
```
Hi @tomsozolins!

The README.md file was updated without use the README.md.gotmpl file. The result is more different that the expected. Do you can to try again? 

Maybe the error is in the directory mounted in make gen-docs command. Try using the helm-docs binary direct, please.

tomsozolins committed:
fix(zabbix): adjust zabbixproxy extraVolumeClaimTemplate

aeciopires commented on README.md:
```diff
@@ -1,6 +1,6 @@
 # Helm Chart For Zabbix.
 
-[![CircleCI](https://circleci.com/gh/cetic/helm-zabbix.svg?style=svg)](https://circleci.com/gh/cetic/helm-zabbix/tree/master) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0) ![version](https://img.shields.io/github/tag/cetic/helm-zabbix.svg?label=release) ![Version: 3.1.1](https://img.shields.io/badge/Version-3.1.1-informational ...
+[![CircleCI](https://circleci.com/gh/cetic/helm-zabbix.svg?style=svg)](https://circleci.com/gh/cetic/helm-zabbix/tree/master) [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0) ![version](https://img.shields.io/github/tag/cetic/helm-zabbix.svg?label=release) ![Version: 3.1.7](https://img.shields.io/badge/Version-3.1.7-informational ...
```
Hi @tomsozolins! Here it should be 3.1.2 instead 3.1.7.

tomsozolins committed:
docs(zabbix): update version

aeciopires reviewed (approved):
Great job @tomsozolins!

Thank you again for your contribution.

tomsozolins commented:
@aeciopires Found an issue with the addition to VolumeClaimTemplate now there is error:
executing "zabbix/templates/statefulset-zabbix-proxy.yaml" at <.Values.zabbixproxy.extraVolumeClaimTemplate>: can't evaluate field Values in type []interface {}

tomsozolins commented:
> @aeciopires Found an issue with the addition to VolumeClaimTemplate now there is error: executing "zabbix/templates/statefulset-zabbix-proxy.yaml" at <.Values.zabbixproxy.extraVolumeClaimTemplate>: can't evaluate field Values in type []interface {}

Should it be like this?
{{- with .Values.zabbixproxy.extraVolumeClaimTemplate }}
  volumeClaimTemplates:
    {{- toYaml . | nindent 4 }}
  {{- end }}

tomsozolins committed:
fix(zabbixproxy): yaml conversion in VolumeClaimTemplate

aeciopires commented:
I understood the problem @tomsozolins. Works with the new config?

aeciopires commented:
Great. Thanks for hotfix. I will test in the next days and will make the merge and publish the version.

Feel free for new improvements.
